v-next/hardhat/src/internal/builtin-plugins/coverage/hook-handlers/network.ts

```diff
@@ -0,0 +1,53 @@
+import type { CoverageManager } from "../types.js";
+import type { HookContext, NetworkHooks } from "hardhat/types/hooks";
+import type { ChainType, NetworkConnection } from "hardhat/types/network";
+
+import { EdrProvider } from "../../network-manager/edr/edr-provider.js";
+
+export default async (): Promise<Partial<NetworkHooks>> => {
+  const handlers: Partial<NetworkHooks> = {
+    async newConnection<ChainTypeT extends ChainType | string>(
+      context: HookContext,
+      next: (
+        nextContext: HookContext,
+      ) => Promise<NetworkConnection<ChainTypeT>>,
+    ) {
+      const connection = await next(context);
+
+      if (context.globalOptions.coverage === true) {
+        if (connection.provider instanceof EdrProvider) {
+          const coverageManager: CoverageManager =
+            // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- We want to access internal coverage manager methods
+            context.coverage as unknown as CoverageManager;
```

Comment:
This does feel a little hacky, too, but the alternative I explored, with making the coverage manager implementation a singleton that can be accessed outside the hre context doesn't feel right either.